src/clients/UBAClient/UBAClientAbstract.ts

```diff
@@ -0,0 +1,139 @@
+import winston from "winston";
+import { RefundRequestWithBlock, UbaFlow } from "../../interfaces";
+import { BigNumber } from "ethers";
+import { UBAFeeSpokeCalculator } from "../../UBAFeeCalculator";
+
+export type RequestValidReturnType = { valid: false; reason: string } | { valid: true };
+export type OpeningBalanceReturnType = { blockNumber: number; spokePoolBalance: BigNumber };
+export type BalancingFeeReturnType = { depositBalancingFee: BigNumber; refundBalancingFee: BigNumber };
+
+/**
+ * UBAClient is a base class for UBA functionality. It provides a common interface for UBA functionality to be implemented on top of or extended.
+ * This class is not intended to be used directly, but rather extended by other classes that implement the abstract methods.
+ */
+export abstract class BaseUBAClient {
+  protected spokeUBAFeeCalculators: { [chainId: number]: { [token: string]: UBAFeeSpokeCalculator } };
+
+  protected constructor(protected readonly chainIdIndices: number[], protected readonly logger?: winston.Logger) {
+    this.spokeUBAFeeCalculators = {};
+  }
+
+  /**
+   * Retrieves the opening balance for a given token on a given chainId at a given block number
+   * @param chainId The chainId to get the opening balance for
+   * @param spokePoolToken The token to get the opening balance for
+   * @param hubPoolBlockNumber The block number to get the opening balance for
+   * @returns The opening balance for the given token on the given chainId at the given block number
+   * @throws If the token cannot be found for the given chainId
+   * @throws If the opening balance cannot be found for the given token on the given chainId at the given block number
+   */
+  public abstract getOpeningBalance(
+    chainId: number,
+    spokePoolToken: string,
+    hubPoolBlockNumber?: number
+  ): OpeningBalanceReturnType;
+
+  /**
+   * Gets the latest block number for a given chainId in the state of the lastest closing block
+   * @param chainId The chainId to get the latest block number for
+   * @returns The latest block number for the given chainId
+   * @note Assumes that the `spoke[...].bundleEndBlocks` are sorted in ascending order
+   */
+  protected abstract resolveClosingBlockNumber(chainId: number, blockNumber: number): number;
```

Comment:
Is the blockNumber argument intended for use in resolving a closing block number in some previous period?

Also wondering if there should be some consideration for a caller-specified follow distance. i.e. if we're trying to resolve a proposed closing block for the current UBA period, we obviously don't want to take the head of a chain that has re-org risk. It's hard to know exactly where in the stack that should be placed, though.

Reply:
This is a fair point. I believe this was part of the previous implementation, and I just re-arranged the structure of this function. It's possible we'd want to include this as a follow-up PR.